Add EX/FA+ Elements to ScreenGameplay

teejusb opened the pull request:
This PR does a couple of things which I will highlight below:
1) Adds an "FA+ Options" row in the first player options screen. In the FA+ game mode, only "Display EX Score" is shown:
These values are saved to a player's profile.
<img width="1111" alt="image" src="https://user-images.githubusercontent.com/5017202/159150847-ea74d66e-cccf-4cf4-9bb7-62ea92f359eb.png">

- "Enabled" Means that the white counts are displayed in step statistics. In the next PR I will make it so that the judgment font chosen will also show white count, but I need to create the graphics for that.
- "Display EX Score" replaces the ITG score with the EX score. This is made obvious by highlighting the score in blue.
<img width="732" alt="image" src="https://user-images.githubusercontent.com/5017202/159150870-717a68f9-a90e-418b-8a17-c33315dac217.png">
<img width="1151" alt="image" src="https://user-images.githubusercontent.com/5017202/159152920-6aabc022-d98e-44a6-a12c-929c073cec15.png">

-----
2) A lot of refactoring surrounding EX/White count calculations
- Added `BGAnimations/ScreenGameplay overlay/TrackExScoreJudgments.lua` - This is used specifically to keep track of the EX Score values/judgments so everything that depends on it will fetch values from there. This actor broadcasts whenever a change has happened and is consumed by `BGAnimations/ScreenGameplay overlay/WhoIsCurrentlyWinning.lua` and `BGAnimations/ScreenGameplay underlay/PerPlayer/Score.lua`

- Modified the helper function `CalculateExScore` to accept the table created from above. The "totals" (taps/holds/rolls) are now computer from `StepsOrTrail` instead of `PlayerStageStats` since it seems like the latter displayed 0s in gameplay. This still seems to work correctly on `ScreenEvaluation` which is evident from the EX scores displayed in `ScreenGameplay` matches up with what's displayed in `ScreenEvaluation`.

- Added a helper function `IsW0Judgment`. This function determines if a TNS falls within the W0 window. This just helps in making it so we didn't need to duplicate the logic.

------
Briefly tested in:
1) 1 Player ITG Mode
2) 1 Player Course Mode
3) 1 Player FA+ Mode
4) 2 Player ITG Mode
5) 2 Player FA+ Mode

teejusb committed:
Add EX/FA+ Elements to ScreenGameplay

natano commented on BGAnimations/ScreenGameplay overlay/TrackExScoreJudgments.lua:
```diff
@@ -0,0 +1,118 @@
+	end,
+	OffCommand=function(self)
+		local storage = SL[ToEnumShortString(player)].Stages.Stats[SL.Global.Stages.PlayedThisGame + 1]
+		storage.ex_counts = ex_counts
```
Can we do this in the `OnCommand` instead? That way we wouldn't have to pass the counts into `CalculateExScore()`. because the counts are always available in the same location.

teejusb replied:
Done.

natano commented on Languages/de.ini:
```diff
@@ -901,6 +903,9 @@ Dark=Dunkel
 Darkest=Dunkelsten
 
+# FaPlus
+Enabled=Ermöglicht
```
"FA+-Fenster anzeigen"

teejusb replied:
Done

natano commented on Languages/de.ini:
```diff
@@ -790,6 +791,7 @@ ComboFont=Wähl die Schriftart für Combo. Diese Schriftart wird auch für den M
 BackgroundFilter=Verstecke die Unterseite des Spielfeld.
 MusicRate=Ändern die Tempo der Musik.
 Stepchart=Wähle ein Schwierigkeit.
+FaPlus=Wählen Sie, ob Sie das FA+-Fenster emulieren und/oder die EX-Wertung im Spiel anzeigen möchten.
```
"Wähle, ob du das FA+-Fenster emulieren und/oder die EX-Wertung im Spiel anzeigen möchten."

teejusb replied:
Done

natano commented on metrics.ini:
```diff
@@ -677,14 +677,20 @@ LineGameplayExtras="lua,CustomOptionRow('GameplayExtras')"
+LineMeasureCounter="lua,CustomOptionRow('MeasureCounter')"
+LineMeasureCounterOptions="lua,CustomOptionRow('MeasureCounterOptions')"
+LineTimingWindows="lua,CustomOptionRow('TimingWindows')"
+LineFaPlus="lua,CustomOptionRow('FaPlus')"
```
I think this should be in the list for the first page.

teejusb replied:
Done

natano commented on Languages/en.ini:
```diff
@@ -926,6 +928,10 @@ Dark=Dark
 Darkest=Darkest
 
+# FaPlus
+Enabled=Enabled
```
Please find a better name.

teejusb replied:
Renamed to `ShowFaPlusWindow=Display FA+ Window`

natano commented on BGAnimations/ScreenGameplay overlay/TrackExScoreJudgments.lua:
```diff
@@ -0,0 +1,118 @@
+local valid_hns = {
+    LetGo = true,
+    Held = true
+}
```
I think we don't need this; all hold note scores are valid.

natano commented on BGAnimations/ScreenGameplay overlay/TrackExScoreJudgments.lua:
```diff
@@ -0,0 +1,118 @@
+return Def.Actor{
+	JudgmentMessageCommand=function(self, params)
+		if params.Player ~= player then return end
+        local count_updated = false
```
inconsistent indentation

teejusb replied:
Oh what the. There were mixed spaces/tabs. Changed to tabs to be consistent with most of SL for now.

natano commented on BGAnimations/ScreenGameplay overlay/TrackExScoreJudgments.lua:
```diff
@@ -0,0 +1,118 @@
+                end
+
+                if not stats:GetFailed() then
+                    ex_counts[adjusted_TNS] = ex_counts[adjusted_TNS] + 1
```
No check for `valid_tns` here?

teejusb replied:
Woops good call! Added.

natano commented on BGAnimations/ScreenGameplay overlay/WhoIsCurrentlyWinning.lua:
```diff
@@ -21,13 +27,29 @@ return Def.Actor{
+			end
+			self:queuecommand("Winning")
+		end
+	end,
```
Maybe don't even add ExCountsChangedMessageCommand in ITG mode and vice versa.

teejusb replied:
Both MessageCommands are relevant in both modes:

You can play ITG mode with EX scoring enabled or disabled
You can play FA+ mode with EX scoring enabled or disabled.

natano commented on BGAnimations/ScreenGameplay underlay/PerPlayer/StepStatistics/TapNoteJudgments.lua:
```diff
@@ -19,18 +23,43 @@ local TNS = {
+			TNS.Names[#TNS.Names+1] = THEME:GetString("TapNoteScore", judgment)
+			TNS.Colors[#TNS.Colors+1] = SL.JudgmentColors["ITG"][5]
+		end
+	else
```
Not sure I get this part. Where is TNS.Names for W0 set?

natano commented:
Reminder: Blue fantastic doesn't increment when using autoplay. That probably hints at some underlying issue.

teejusb commented:
Okay I think I've addressed all the comments:

At a high level the changes are:
1) Remove `ex_counts` argument from `CalculateExScore` and have the function fetch this from storage.
2) Renamed `Enabled` to `ShowFaPlusWindow` and fixed verbiage across langauges
3) Bug with blue fantastic not incremented fixed. It was a typo that I made after creating the PR.

natano reviewed (approved):
LGTM!
